### ISSUE TITLE
Use reflection for mocked objects injection

### ISSUE THREAD

**Issue by fesor** [peridot-php/peridot-prophecy-plugin#4]
PhpSpec has it's great sugar for injection mocked objects. Is it possible to implement such thing in Peridot?

For example

``` php
<?php
describe('Bird', function() {

    beforeEach(function(Wings $wings) {
        $this->$bird = new Bird($wings);
    });

    it('should fly', function(Wings $wings) {
        $wings->swing()->shouldBeCalled();
        $this->bird->fly();
    });
});
```

**Comment by fesor** [comment]
Duplicate of #1.

**Comment by brianium** [comment]
I was waiting for someone else to take interest in this :) As of Peridot 1.15.0 it is possible for plugins to provide arguments to spec functions - so we could pass a prophecy object as an argument.

I think the complexity would come from the fact that the result of `prophesize` is an instance of `ObjectProphecy` - and I don't think that can be an instance of `Wing`. PhpSpec does a lot of work to make this happen, and though cool, it is very slow. I wonder if we could get around it with some sort of simple proxy object. I will investigate some options.

**Comment by fesor** [comment]
Prophet will return ObjectProphecy without doubler only if mocking class/interface [doesn't exists](https://github.com/phpspec/prophecy/blob/master/src/Prophecy/Prophet.php#L83). If class exists, you can get doubled object with `reveal()` method and inject it into callback arguments list. Need to check this to be sure.

**Comment by brianium** [comment]
I believe you are correct, but the double will not have access to spies or method prophecies correct? The example above would not be possible. `$wing` would not have access to `->shouldBeCalled()` - unless I am missing something?

**Comment by fesor** [comment]
Yes, you are right... The only possible way to call 'willReturn' is

``` php
function (Wings $wings) {
    $wings->getProphet()->swing()->willReturn('fly like an eagle');
}
```

Or...

``` php
<?php
describe('Bird', function() {

    beforeEach(function(Wings $wings) {
        // $wings is actually ObjectProphecy, or wrapped one
        $this->$bird = new Bird($wings->reveal()); // or ->unwrap() or ->getMock()
    });

    it('should fly', function(Wings $wings) {
        $wings->swing()->shouldBeCalled();
        $this->bird->fly();
    });
});
```

Unless testing class is wrapped with some proxy, it is very complicated to make this the same way as in phpspec.

**Comment by brianium** [comment]
Challenge accepted! I'll spend a little time playing with a lightweight proxy implementation.
